docker/docker_image_dest.go

```diff
@@ -351,8 +358,24 @@ func (d *dockerImageDestination) TryReusingBlob(ctx context.Context, info types.
 				continue
 			}
 		}
-		cache.RecordKnownLocation(d.ref.Transport(), bicTransportScope(d.ref), candidate.Digest, newBICLocationReference(d.ref))
-		return true, types.BlobInfo{Digest: candidate.Digest, Size: size}, nil
+		cache.RecordKnownLocation(d.ref.Transport(), bicTransportScope(d.ref), candidate.Digest, candidate.CompressorName, newBICLocationReference(d.ref))
+
+		var compressionOperation types.LayerCompression
+		var compressionAlgorithm *compression.Algorithm
+		if candidate.CompressorName != "" {
+			compressionOperation = types.Compress
+			algorithm, err := compression.AlgorithmByName(candidate.CompressorName)
+			if err != nil {
+				logrus.Debugf("... unknown compressor %q", candidate.CompressorName)
+				return false, types.BlobInfo{}, nil
+			}
+			compressionAlgorithm = &algorithm
+		}
+		mediaType, err := mediatypes.AnyUpdatedMIMEType(info.MediaType, types.BlobInfo{CompressionOperation: compressionOperation, CompressionAlgorithm: compressionAlgorithm})
```

Comment:
I haven’t read the code in full at all, (and I apologize if I’m missing the whole point), but this feels misplaced at a first glance.

Originally I was expecting the cache to record the raw MIME type value, `TryReusingBlob` to report the raw MIME type value, and the _generic code_ deal with the concept of manifest-format-specific MIME types. It’s a very good insight that “MIME type of a blob” is not really a real concept with the v2s2/OCI alternatives, still, the transports should to an extent not need to worry about that.

Especially inventing a ~random MIME type based on the input (not output) format, and the compression mechanism seems superfluous — if the MIME type is really that bogus, maybe we should be deprecating/removing it completely? OTOH that seems way too much.

(I have no idea what the right answer is — at this point I’m just thinking that the difficulty of changing APIs makes it important to get this right if possible.)

---

This doesn’t handle uncompressed formats AFAICS.

Reply:
An earlier version of the patch recorded the MIME types.  For cases where we ended up doing format conversions, that required adding a way to map from "schema 2's name for a gzip-compressed layer" in a BICReplacementCandidate to "OCI's name for a gzip-compressed layer" in copy, or in TryReusingBlob(), where the manifest's type isn't currently exposed.  I found that recording just the compressor names in the cache and passing MediaType values across the existing interfaces let me avoid adding that.

For comparison, the WIP before that change is https://github.com/nalind/image/tree/blob-info-cache-media-types-1.  It didn't move mappings from the `manifest` package to `internal/mediatypes`, so the diff stats aren't going to be very useful.  I wasn't happy with having so many ad-hoc lookup tables scattered across the code base.

The intent is to use `MTSUncompressed` as the "compressor" for uncompressed formats.  It's possible that part of that is missing and causing the test failures, but I'm thinking that if the blob cache cares more about the MediaType field than it used to, that some of those are due to the schema 1 manifests not supplying values there.

Reply:
> For comparison, the WIP before that change is https://github.com/nalind/image/tree/blob-info-cache-media-types-1. 

Thanks, I’ll take a look.

>  I wasn't happy with having so many ad-hoc lookup tables scattered across the code base.

Very vaguely I guess that it should be possible to build a single master table of (manifest format, blob semantics, compression format) → MIME type in a way that allows the relevant lookups (“what is this type with compression changed? what is the type for a different manifest?” etc.) Building that for O(1) lookups and no repetition would be some work, but the tables are currently very small — linear lookups would might well be perfectly OK.

(The existence of `CompressionOperation` inside `BlobInfo` is conceptually wrong — it is a “make this edit” temporary state of the copy process just for `ManifestUpdateOptions`, just something we didn’t catch in time. Making this a field that is fair game to be manipulated by the transports complicates the analysis quite a bit, or requires setting and documenting detailed restrictive expectations.)

---

> The intent is to use `MTSUncompressed` as the "compressor" for uncompressed formats. It's possible that part of that is missing and causing the test failures, but I'm thinking that if the blob cache cares more about the MediaType field than it used to, that some of those are due to the schema 1 manifests not supplying values there.

That could work, but the way I’m reading the code (untested), `mediatypes.UpdatedMIMEType` only does a table lookup if the `CompressionOperation` is not `PreserveOriginal`, and this code only sets `CompressionOperation` to `Compress` if the compressor is non-empty. A quick `if candidate.CompressorName != "" { … } else { compressionOperation = types.Decompress }` might be all it takes.

Reply:
I think you're right about the cases where we want to change the type of compression being used for a layer.

Reply:
Modified `UpdatedMIMEType()` to treat `PreserveOriginal` with a non-empty compressor name differently, to update the media type.  Setting the operation to `Decompress` when there's no compressor set could have us decompress something that's already compressed when we don't intend that, and  I think `PreserveOriginal` with no compressor specified happens a lot, so holding off on that, at least for now.